### PR TITLE
Adding automated version check on server

### PR DESCRIPTION
**Pull request by vitarb**
This change adds a version checker deamon that will upon startup and every 24 hours collect version guidance from the version-info service and will store it in the database.

This feature will be enabled by default. It's possible to turn it off by either setting `TEMPORAL_DISABLE_VERSION_CHECK` env var or by using `frontend.enableServerVersionCheck` dynamic config property.

### CODE REVIEW

File: service/frontend/versionChecker.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	if vc.config.EnableServerVersionCheck() {
+		go vc.versionCheckLoop()
+	}
+}
+
+func (vc *VersionChecker) Stop() {
+	if vc.config.EnableServerVersionCheck() {
+		close(vc.shutdownChan)
+	}
+}
```
Comment (wxing1292) [inline service/frontend/versionChecker.go:46]: what if the start / stop function are called multiple times?

usually the start will only be called by initialization logic
there is no guarantee for stop

File: service/frontend/versionChecker.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+func (vc *VersionChecker) performVersionCheck() {
+	sw := vc.metricsScope.StartTimer(metrics.VersionCheckLatency)
+	defer sw.Stop()
+	req, err := vc.createVersionCheckRequest()
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
```
Comment (wxing1292) [inline service/frontend/versionChecker.go:68]: emit a error log?

File: service/frontend/versionChecker.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	resp, err := vc.getVersionInfo(req)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
```
Comment (wxing1292) [inline service/frontend/versionChecker.go:73]: emit a error log?

File: service/frontend/versionChecker.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	err = vc.saveVersionInfo(resp)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
```
Comment (wxing1292) [inline service/frontend/versionChecker.go:78]: emit a error log?

File: service/frontend/versionChecker.go (hunk 2 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:68]: We don't want to log intentionally. If this feature is not working due to lack of network connectivity for example we don't want to cause any concerns for our users.

File: service/frontend/versionChecker.go (hunk 1 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:46]: As agreed up on offline, we'll ignore this for now and will have a broader discussion later.

Comment (wxing1292) [reply service/frontend/versionChecker.go:46]: seems that all `start` / `stop` functions are now *not* checking at most once execution.

let me schedule a meeting to find out why and possibly add the at most once check back

Comment (wxing1292) [reply service/frontend/versionChecker.go:46]: just FYI, if the stop is called >= 2 times, it will panic?

Comment (wxing1292) [reply service/frontend/versionChecker.go:46]: how about adding a https://golang.org/pkg/sync/#Once.Do for now

File: common/resource/resource.go [hunk 5]
```diff
@@ -94,6 +94,7 @@ type (
 		// persistence clients
 
 		GetMetadataManager() persistence.MetadataManager
+		GetClusterMetadataManager() persistence.ClusterMetadataManager
```
Comment (alexshtin) [inline common/resource/resource.go:97]: Why not just to add it to version checker?

File: service/frontend/versionChecker.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	resp, err := vc.getVersionInfo(req)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
+		return
+	}
+	err = vc.saveVersionInfo(resp)
+	if err != nil {
+		vc.metricsScope.IncCounter(metrics.VersionCheckFailedCount)
```
Comment (alexshtin) [inline service/frontend/versionChecker.go:95]: Do you want to distinguish network and DB failures in metrics?

File: service/frontend/versionChecker.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+			Severity: int32(alert.Severity),
+		})
+	}
+	return result
+}
+
+func convertReleaseInfo(releaseInfo check.ReleaseInfo) *persistenceblobs.ReleaseInfo {
+	return &persistenceblobs.ReleaseInfo{
+		Version:     releaseInfo.Version,
+		ReleaseTime: timestamp.TimestampFromTime(time.Unix(releaseInfo.ReleaseDate, 0)).ToProto(),
```
Comment (alexshtin) [inline service/frontend/versionChecker.go]: Please fix this in proto file: `release_time` should have `[(gogoproto.stdtime) = true]` suffix (check other fields). Then here, in go, `ReleaseTime` will be just `*time.Time` and you won't need this conversion.

File: service/frontend/versionChecker.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+		Alerts:       convertAlerts(resp.Alerts),
+	}
+}
+
+func convertAlerts(alerts []check.Alert) []*persistenceblobs.Alert {
+	var result []*persistenceblobs.Alert
+	for _, alert := range alerts {
+		result = append(result, &persistenceblobs.Alert{
+			Message:  alert.Message,
+			Severity: int32(alert.Severity),
```
Comment (alexshtin) [inline service/frontend/versionChecker.go:156]: Change `alert.Severity` to `int32` in first place.

File: service/frontend/versionChecker.go [hunk 9]
```diff
@@ -127,11 +138,13 @@ func (vc *VersionChecker) saveVersionInfo(resp *check.VersionCheckResponse) erro
...
 	return &persistenceblobs.VersionInfo{
-		Current:      convertReleaseInfo(resp.Current),
-		Recommended:  convertReleaseInfo(resp.Recommended),
-		Instructions: resp.Instructions,
-		Alerts:       convertAlerts(resp.Alerts),
+		Current:        convertReleaseInfo(resp.Current),
+		Recommended:    convertReleaseInfo(resp.Recommended),
+		Instructions:   resp.Instructions,
+		Alerts:         convertAlerts(resp.Alerts),
+		LastUpdateTime: &currentTime,
```
Comment (wxing1292) [inline service/frontend/versionChecker.go:147]: timestamp.TimePtr?

**Review by wxing1292 — APPROVED** [review]
plz provide some tests before landing

File: service/frontend/versionChecker.go [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+		Arch:      runtime.GOARCH,
+		OS:        runtime.GOOS,
+		DB:        clusterMetadataManager.GetName(),
+		ClusterID: metadata.ClusterId,
+		Timestamp: time.Now().UnixNano(),
+	}, nil
+}
+
+func (vc *VersionChecker) getVersionInfo(req *check.VersionCheckRequest) (*check.VersionCheckResponse, error) {
+	return check.NewCaller().Call(req)
```
Comment (alexshtin) [inline service/frontend/versionChecker.go:119]: So, this will never panic?

File: service/frontend/versionChecker.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	shutdownChan chan struct{}
+	metricsScope metrics.Scope
+	startOnce    sync.Once
+	stopOnce     sync.Once
+}
+
+func NewVersionChecker(
+	resource resource.Resource,
+	params *resource.BootstrapParams,
+	config *Config,
```
Comment (alexshtin) [inline service/frontend/versionChecker.go:33]: and align.

File: service/frontend/versionChecker.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+		return err
+	}
+	if !saved {
+		return serviceerror.NewInternal("version info update hasn't been applied")
+	}
+	return nil
+}
+
+func toVersionInfo(resp *check.VersionCheckResponse) *persistenceblobs.VersionInfo {
+	currentTime := time.Now()
```
Comment (alexshtin) [inline service/frontend/versionChecker.go:141]: I would suggest to always add `.UTC()` to `time.Now()`. It probably doesn't matter here but it is a good habit.

File: service/frontend/versionChecker.go (hunk 10 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:119]: It shouldn't, if it doesn't it's fine too as we run in sync.Once

File: service/frontend/versionChecker.go (hunk 12 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:141]: Done

File: service/frontend/versionChecker.go (hunk 6 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:95]: Possibly, we can add VersionCheckRequestFailedCount or something along those lines.

File: service/frontend/versionChecker.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/version/check"
```
[the remark this answers is not in the document]
Comment (vitarb) [reply service/frontend/versionChecker.go:15]: I've tried running `make goimports` but it reformatted a bunch of unrelated files...

File: service/frontend/versionChecker.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
+package frontend
```
[the remark this answers is not in the document]
Comment (vitarb) [reply service/frontend/versionChecker.go:1]: It's in different package...

File: go.sum [hunk 15]
```diff
@@ -359,6 +359,10 @@ go.temporal.io/api v1.0.0 h1:mWtvS+5ENYvG4ZPZ/4/bxCj4j3gIF4D05C2GVrhLpjc=
 go.temporal.io/api v1.0.0/go.mod h1:AgbKINgV3KR9SlTH8nQRsNadVbxVI+/LnZ1uFModMIA=
 go.temporal.io/sdk v1.0.0 h1:Cfrr/RkcoGu+B/vpOII1D7xv4rhRHa5eYQStnuASS7k=
 go.temporal.io/sdk v1.0.0/go.mod h1:mtuFg8AafvaXeAE3gwBjZZN05uwTGjzVf1qkXmq4hpM=
+go.temporal.io/version v0.0.0-20201005024203-3856dc1d8c33 h1:V0VvsblOgBfP3NRr+iH5VO61BfBhy+gXmuC4MqHiJ7I=
+go.temporal.io/version v0.0.0-20201005024203-3856dc1d8c33/go.mod h1:UA9S8/1LaKYae6TyD9NaPMJTZb911JcbqghI2CBSP78=
+go.temporal.io/version v0.0.0-20201009034513-3988ea99117a h1:3U3YqWRMoxTwjbXkwWPtbQuLn+yav9uEwYwfV5v9gPE=
+go.temporal.io/version v0.0.0-20201009034513-3988ea99117a/go.mod h1:UA9S8/1LaKYae6TyD9NaPMJTZb911JcbqghI2CBSP78=
```
[the remark this answers is not in the document]
Comment (vitarb) [reply go.sum:365]: Done

File: service/frontend/service.go [hunk 16]
```diff
@@ -178,9 +183,10 @@ type Service struct {
 	config *Config
 	params *resource.BootstrapParams
 
-	handler      Handler
-	adminHandler *AdminHandler
-	server       *grpc.Server
+	handler        Handler
+	adminHandler   *AdminHandler
+	versionChecker *VersionChecker
```
[the remark this answers is not in the document]
Comment (vitarb) [reply service/frontend/service.go:188]: Discussed offline. Best place for now.

File: service/frontend/versionChecker.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (vitarb) [reply service/frontend/versionChecker.go:1]: Discussed offline, we'll rename the old one later.

File: common/resource/resource.go (hunk 5 above)
Comment (vitarb) [reply common/resource/resource.go:97]: This was left out before, just fixing it here.

File: service/frontend/versionChecker.go (hunk 8 above)
Comment (vitarb) [reply service/frontend/versionChecker.go:156]: Ok

File: service/frontend/versionChecker.go (hunk 7 above)
Comment (vitarb) [reply service/frontend/versionChecker.go]: Done

**Review by alexshtin — APPROVED** [review]

**Pull request merged**